src/lib/storage/dictionary_segment.cpp

```diff
@@ -1,95 +1,157 @@
 #include "dictionary_segment.hpp"
 
+#include <map>
+#include <set>
+#include "fixed_width_integer_vector.hpp"
+#include "type_cast.hpp"
 #include "utils/assert.hpp"
+#include "value_segment.hpp"
 
 namespace opossum {
 
+std::shared_ptr<AbstractAttributeVector> get_attribute_vector(size_t vector_size, size_t size) {
+  auto bits_needed = std::bit_width(vector_size - 1);
+  Assert(bits_needed <= 32, "Too many values in dictionary, cant use more than 32 bits!");
+  if (bits_needed <= 8) {
+    return std::make_shared<FixedWidthIntegerVector<uint8_t>>(size);
+  } else if (bits_needed <= 16) {
+    return std::make_shared<FixedWidthIntegerVector<uint16_t>>(size);
+  } else {
+    return std::make_shared<FixedWidthIntegerVector<uint32_t>>(size);
+  }
+}
+
 template <typename T>
 DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
-  // Implementation goes here
+  // retrieve value segment
+  const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(abstract_segment);
+  Assert(value_segment, "Can't encode abstract segment, because it is no value segment.");
```

Comment:
The abstract segment might be a value segment, just not of type T, should we reflect this in this Assert-message or assert it seperately?